src/Controller/ControllerFactory.php

```diff
@@ -174,8 +174,16 @@ protected function getActionArgs(Closure $action, array $passedParams): array
 
             // Check for dependency injection for classes
             if ($type instanceof ReflectionNamedType && !$type->isBuiltin()) {
-                if ($this->container->has($type->getName())) {
-                    $resolved[] = $this->container->get($type->getName());
+                $typeName = $type->getName();
+                if ($this->container->has($typeName)) {
+                    $resolved[] = $this->container->get($typeName);
+                    continue;
+                }
+
+                // Use passedParams as a source of typed dependencies.
+                // The accepted types for passedParams was never defined and userland code relies on that.
+                if ($passedParams && is_object($passedParams[0]) && $passedParams[0] instanceof $typeName) {
```

Comment:
Why check for objects only? Scalar types could also be pre converted in passed args.

Reply:
Because the guard condition above is for non-builtin types. The scalar values are all builtin.

Reply:
Ah sorry, I missed the `!$type->isBuiltin()` check above.